### PR TITLE
feat: .redocly.yaml config changes

### PR DESCRIPTION
**Pull request by andriyl**
## What/Why/How?
pitch:  .redocly.yaml config changes
- [x] config format trasformer
- [x] lint merger
- [ ] latest version trasformer
- [x] cover with tests

## Check yourself
- [x] Code is linted
- [ ] Tested with redoc/reference-docs/workflows
- [x] All new/updated code is covered with tests

## Security
- [x] Security impact of change has been considered
- [x] Code follows company security practices and guidelines

### CODE REVIEW

File: packages/core/src/config/load.ts [hunk 1]
```diff
@@ -61,6 +50,38 @@ export async function loadConfig(configPath?: string, customExtends?: string[]):
...
+  const keysToChange = ['apiDefinitions', 'referenceDocs'];
+
+  for (const [key, value] of Object.entries(rawConfig)) {
+    if (key === 'apiDefinitions') {
+      config.apis = {};
+      for (const [apiName, apiPath] of Object.entries(value)) {
+        config.apis[apiName+'@latest'] = { 'root': apiPath };
+      }
+    }
+    if (key === 'referenceDocs') { config['features.openapi'] = value; }
```
Comment (tatomyr) [inline packages/core/src/config/load.ts:66]: Could collide with an existing 'features.openapi' field

Comment (RomanHotsiy) [reply packages/core/src/config/load.ts:66]: what is existing 'features.openapi'?

Comment (tatomyr) [reply packages/core/src/config/load.ts:66]: I mean, **features.openapi** could be already defined in **redocly.yaml**. And here we are rewriting it with the value of **referenceDocs**.

Comment (andriyl) [reply packages/core/src/config/load.ts]: Outdated.

File: packages/core/src/config/load.ts [hunk 2]
```diff
@@ -61,6 +48,55 @@ export async function loadConfig(configPath?: string, customExtends?: string[]):
...
+  if ((rawConfig as RawConfig).apis && (rawConfig as DeprecatedRawConfig).apiDefinitions ||
+    (rawConfig as RawConfig)['features.openapi'] && (rawConfig as DeprecatedRawConfig).referenceDocs
+  ) {
+    throw new Error('Do not use old & new config syntax simultaneously');
+  }
+  const { apiDefinitions, referenceDocs, ...rest } = rawConfig as DeprecatedRawConfig & RawConfig;
+  if (apiDefinitions || referenceDocs) {
+    // TODO: add link to docs.
+    // TODO: show warning without throwing error.
+    throw new Error('apiDefinitions & referenceDocs fields are deprecated. Use apis & features.openapi instead (see  docs: /.//) ');
```
Comment (tatomyr) [inline packages/core/src/config/load.ts:78]: Please put a meaningful comment. 
@adamaltman, @RomanHotsiy  could you suggest one?

**Review by tatomyr — CHANGES_REQUESTED** [review]

File: packages/core/src/config/load.ts (hunk 2 above)
Comment (adamaltman) [reply packages/core/src/config/load.ts:78]: Well, I think this doesn't make sense:

```
  if (apiDefinitions || referenceDocs) {
```

Each should have its own warning (not error). The warning message can be:

- The 'apiDefinitions' field is deprecated. Use 'apis' instead.
- The 'referenceDocs' field is deprecated. Use 'features.openapi' instead.

File: packages/cli/src/commands/lint.ts [hunk 3]
```diff
@@ -18,6 +18,24 @@ import { Totals } from '../types';
 import { blue, gray } from 'colorette';
 import { performance } from 'perf_hooks';
 
+export function mergeLintConfigs(entrypoint: any, config: any) {
+  if (!entrypoint.alias) return config;
+  let mergedLint = config.apis[entrypoint.alias]?.lint || {};
```
Comment (adamaltman) [inline packages/cli/src/commands/lint.ts:23]: I'm not totally clear here, but what if the `config` is empty or doesn't exist?

File: packages/core/src/config/load.ts (hunk 2 above)
Comment (andriyl) [reply packages/core/src/config/load.ts:78]: @adamaltman thx

File: packages/core/src/config/load.ts [hunk 4]
```diff
@@ -61,6 +48,55 @@ export async function loadConfig(configPath?: string, customExtends?: string[]):
...
+    apis[apiName] = { root: apiPath };
+  }
+  return apis;
+}
+
+function decorateWithLatest(rawConfig: RawConfig): RawConfig {
+  let apis: Record<string, Api> = {};
+  for (const [key, value] of Object.entries(rawConfig.apis || {})) {
+    const [name, version] = key.split('@');
+    apis[`${name}@${version || 'latest'}`] = value;
```
[the remark this answers is not in the document]
Comment (andriyl) [reply packages/core/src/config/load.ts]: Outdated, removed `latest version transformer for now.

File: packages/cli/src/commands/lint.ts (hunk 3 above)
Comment (andriyl) [reply packages/cli/src/commands/lint.ts:25]: @adamaltman that's ok, we always have root lint with default plugins inside - will merge successfully anyway.

File: packages/cli/src/commands/lint.ts [hunk 5]
```diff
@@ -52,10 +73,10 @@ export async function handleLint(
   for (const entryPoint of entrypoints) {
     try {
       const startedAt = performance.now();
-      process.stderr.write(gray(`validating ${entryPoint.replace(process.cwd(), '')}...\n`));
+      process.stderr.write(gray(`validating ${entryPoint.path.replace(process.cwd(), '')}...\n`));
       const results = await lint({
-        ref: entryPoint,
-        config,
+        ref: entryPoint.path,
+        config: mergeLintConfigs(entryPoint, config)
```
[the remark this answers is not in the document]
Comment (andriyl) [reply packages/cli/src/commands/lint.ts]: done

File: packages/cli/src/commands/lint.ts [hunk 6]
```diff
@@ -18,6 +18,24 @@ import { Totals } from '../types';
 import { blue, gray } from 'colorette';
 import { performance } from 'perf_hooks';
 
+export function mergeLintConfigs(entrypoint: any, config: any) {
```
[the remark this answers is not in the document]
Comment (andriyl) [reply packages/cli/src/commands/lint.ts]: done

File: packages/cli/src/utils.ts [hunk 7]
```diff
@@ -18,16 +18,18 @@ import {
...
-  }
+    ? Object.keys(apis).map((fileName) => ({
+        path: resolve(getConfigDirectory(config), apis[fileName].root),
+        alias: fileName
+      }))
+    : (await expandGlobsInEntrypoints(argsEntrypoints!, config)).map(entrypoint => ({ path: entrypoint }));
+  // if (!isNotEmptyArray(res)) {
+  //   process.stderr.write('error: missing required argument `entrypoints`.\n');
+  //   process.exit(1);
+  // }
```
[the remark this answers is not in the document]
Comment (andriyl) [reply packages/cli/src/utils.ts]: done

File: packages/core/src/config/load.ts (hunk 2 above)
Comment (andriyl) [reply packages/core/src/config/load.ts]: done

File: packages/cli/src/commands/join.ts [hunk 8]
```diff
@@ -56,7 +56,7 @@ packageVersion: string
   const entrypoints = await getFallbackEntryPointsOrExit(argv.entrypoints, config);
   const externalRefResolver = new BaseResolver(config.resolve);
   const documents = await Promise.all(
-    entrypoints.map(ref => externalRefResolver.resolveDocument(null, ref, true) as Promise<Document>)
+    entrypoints.map(ref => externalRefResolver.resolveDocument(null, ref.path, true) as Promise<Document>)
```
Comment (tatomyr) [inline packages/cli/src/commands/join.ts:59]: `entrypoints.map(({path}) => externalRefResolver.resolveDocument(null, path, true) as Promise<Document>)`

File: packages/cli/src/commands/bundle.ts [hunk 9]
```diff
@@ -64,10 +63,7 @@ export async function handleBundle(
             )} configuration by default.\n\n`,
           );
         }
-        const results = await lint({
-          ref: entrypoint,
-          config,
-        });
+        const results = await lint({ ref: path, config });
```
Comment (tatomyr) [inline packages/cli/src/commands/bundle.ts:66]: Merge lint configs?

File: packages/cli/src/commands/bundle.ts [hunk 10]
```diff
@@ -83,22 +79,22 @@ export async function handleBundle(
...
+      process.stderr.write(gray(`bundling ${path}...\n`));
 
       const {
         bundle: result,
         problems,
         ...meta
       } = await bundle({
-        config,
-        ref: entrypoint,
+        config: mergeLintConfigs(config, alias),
```
Comment (tatomyr) [inline packages/cli/src/commands/bundle.ts:89]: Why `bundle` needs the lint config?

File: packages/core/src/config/load.ts [hunk 11]
```diff
@@ -61,6 +50,68 @@ export async function loadConfig(configPath?: string, customExtends?: string[]):
   );
 }
 
+export function mergeLintConfigs(config: Config, entrypointAlias?: string): Config {
```
Comment (tatomyr) [inline packages/core/src/config/load.ts:53]: Rename this

File: packages/cli/src/utils.ts [hunk 12]
```diff
@@ -15,15 +15,17 @@ import {
...
+export async function getFallbackEntryPointsOrExit(argsEntrypoints: string[] | undefined, config: Config): Promise<Entrypoint[]> {
+  const { apis } = config;
+  const shouldFallbackToAllDefinitions = !isNotEmptyArray(argsEntrypoints) && apis && Object.keys(apis).length > 0;
   const res = shouldFallbackToAllDefinitions
-    ? Object.values(apiDefinitions).map((fileName) => resolve(getConfigDirectory(config), fileName))
-    : await expandGlobsInEntrypoints(argsEntrypoints!, config);
-
+    ? Object.keys(apis).map((fileName) => ({
+        path: resolve(getConfigDirectory(config), apis[fileName].root),
+        alias: fileName
```
Comment (tatomyr) [inline packages/cli/src/utils.ts:26]: Add a comma

File: packages/cli/src/utils.ts [hunk 13]
```diff
@@ -35,11 +37,14 @@ function getConfigDirectory(config: Config) {
   return config.configFile ? dirname(config.configFile) : process.cwd();
 }
 
-function isNotEmptyArray(args?: string[]): boolean {
+function isNotEmptyArray(args?: string[] | object[]): boolean {
```
Comment (tatomyr) [inline packages/cli/src/utils.ts:40]: Fix typing

File: packages/core/src/config/load.ts (hunk 11 above)
Comment (tatomyr) [reply packages/core/src/config/load.ts:53]: Introduce a separate function for merging only lint configs

File: packages/core/src/config/load.ts [hunk 14]
```diff
@@ -61,6 +50,68 @@ export async function loadConfig(configPath?: string, customExtends?: string[]):
...
+  const apis: Record<string, Api> = {};
+  for (const [apiName, apiPath] of Object.entries(apiDefinitions)) {
+    apis[apiName] = { root: apiPath };
+  }
+  return apis;
+}
+
+function transformConfig(rawConfig: DeprecatedRawConfig | RawConfig): RawConfig {
+  if ((rawConfig as RawConfig).apis && (rawConfig as DeprecatedRawConfig).apiDefinitions ||
+    (rawConfig as RawConfig)['features.openapi'] && (rawConfig as DeprecatedRawConfig).referenceDocs
```
Comment (tatomyr) [inline packages/core/src/config/load.ts:80]: Maybe use a separate boolean const?

Comment (tatomyr) [reply packages/core/src/config/load.ts:80]: Use separate errors for different fields?

File: packages/core/src/config/load.ts [hunk 15]
```diff
@@ -61,6 +50,68 @@ export async function loadConfig(configPath?: string, customExtends?: string[]):
...
+    );
+  }
+  return {
+    'features.openapi': referenceDocs,
+    apis: transformApiDefinitionsToApis(apiDefinitions),
+    ...rest
+  };
+}
+
+export async function getConfig(path?: string) {
```
Comment (tatomyr) [inline packages/core/src/config/load.ts:102]:
```suggestion
export async function getConfig(path: string = findConfig()) {
```

File: packages/core/src/config/load.ts [hunk 16]
```diff
@@ -61,6 +50,68 @@ export async function loadConfig(configPath?: string, customExtends?: string[]):
...
+  let rawConfig: RawConfig = {};
+  const configPath = path || findConfig();
+  if (configPath) {
+    try {
+      rawConfig = (await loadYaml(configPath)) as RawConfig;
```
Comment (tatomyr) [inline packages/core/src/config/load.ts:107]: Return here!

File: __tests__/bundle/bundle-lint-format/.redocly.yaml [hunk 17]
```diff
@@ -1,2 +1,3 @@
-apiDefinitions:
-  main: ./openapi.yaml
+apis:
+  main:
```
Comment (adamaltman) [inline __tests__/bundle/bundle-lint-format/.redocly.yaml:2]: None of the test examples use the `name@version` approach.

File: packages/core/src/config/load.ts [hunk 18]
```diff
@@ -61,6 +49,16 @@ export async function loadConfig(configPath?: string, customExtends?: string[]):
...
+  if (!configPath) return {};
+  try {
+    const rawConfig = (await loadYaml(configPath)) as RawConfig;
+    return transformConfig(rawConfig);
+  } catch (e) {
+    throw new Error(`Error parsing config file at '${configPath}': ${e.message}`);
+  }
+}
+
 function findConfig() {
```
Comment (AyeOne) [inline packages/core/src/config/load.ts:62]: `redocly.yaml` and `redocly.yml` support is missing here

Comment (AyeOne) [reply packages/core/src/config/load.ts:62]: This should work fine:
```
 const supportedConfigNames = ['redocly.yaml', 'redocly.yml', '.redocly.yaml', `.redocly.yml`];
  const configName = supportedConfigNames.find((fileName) => {
    return existsSync(`${definitionDir}/${fileName}`);
  });
  return configName;
  ```

Comment (adamaltman) [reply packages/core/src/config/load.ts:62]: There was this separate PR about that I think: https://github.com/Redocly/openapi-cli/pull/557/files

File: __tests__/bundle/bundle-lint-format/.redocly.yaml (hunk 17 above)
Comment (tatomyr) [reply __tests__/bundle/bundle-lint-format/.redocly.yaml:2]: Anyway, good point. We'll need to write the tests after implementing this functionality.

Comment (RomanHotsiy) [reply __tests__/bundle/bundle-lint-format/.redocly.yaml:2]: This will be needed for push.

File: packages/cli/src/commands/lint.ts [hunk 19]
```diff
@@ -35,27 +36,29 @@ export async function handleLint(
...
-    );
-  }
 
   // TODO: use shared externalRef resolver, blocked by preprocessors now as they can mutate documents
-  for (const entryPoint of entrypoints) {
+  for (const { path, alias } of entrypoints) {
     try {
       const startedAt = performance.now();
-      process.stderr.write(gray(`validating ${entryPoint.replace(process.cwd(), '')}...\n`));
+      const mergedConfig = getMergedConfig(config, alias);
```
Comment (RomanHotsiy) [inline packages/cli/src/commands/lint.ts:50]: How about renaming it to `resolvedConfig`?

File: packages/cli/src/utils.ts [hunk 20]
```diff
@@ -35,11 +37,14 @@ function getConfigDirectory(config: Config) {
   return config.configFile ? dirname(config.configFile) : process.cwd();
 }
 
-function isNotEmptyArray(args?: string[]): boolean {
+function isNotEmptyArray(args?: any[]): boolean {
```
Comment (RomanHotsiy) [inline packages/cli/src/utils.ts:40]: Let it work with any type.

```suggestion
function isNotEmptyArray<T>(args?: T[]): boolean {
```

File: packages/core/src/config/config.ts [hunk 21]
```diff
@@ -624,3 +633,69 @@ function assignExisting<T>(target: Record<string, T>, obj: Record<string, T>) {
...
+export function getMergedLintConfig(
+  config: Config,
+  entrypointAlias?: string
+): LintConfig {
+  const localLint = entrypointAlias 
+    ? config.apis[entrypointAlias]?.lint
+    : {};
+  const mergedLint = {
+    ...config.rawConfig.lint,
+    // Preserve external additions to the lint config:
```
Comment (RomanHotsiy) [inline packages/core/src/config/config.ts:660]: what are those external additions?

**Review by RomanHotsiy — COMMENTED** [review]
@tatomyr this looks good. I left some comments. I would like to take another look before merging.

File: packages/core/src/config/config.ts (hunk 21 above)
Comment (tatomyr) [reply packages/core/src/config/config.ts:660]: The `config.lint` gets modified `loadConfig`. So it contains some additional attributes if I remember correctly.

**Pull request closed without merging**